lib/core/src/Cardano/Wallet.hs

```diff
@@ -1420,7 +1420,12 @@ estimateFeeForDelegation ctx wid = db & \DBLayer{..} -> do
     let action = if isKeyReg then Join pid else RegisterKeyAndJoin pid
     let selectCoins = selectCoinsForDelegationFromUTxO @_ @t @k
             ctx utxo txp minUtxo action
-    estimateFeeForCoinSelection $ Fee . feeBalance <$> selectCoins
+
+    dep <- fmap stakeKeyDeposit $
+        withExceptT ErrSelectForDelegationNoSuchWallet
+            $ readWalletProtocolParameters @ctx @s @k ctx wid
+    estimateFeeForCoinSelection (if isKeyReg then Nothing else Just $ getCoin dep)
```

Comment:
Note that passing this to the `estimateFeeForCoinSelection` function is a bit redundant with the coin selection itself. A coin selection is captured by:

```hs
data CoinSelection = CoinSelection
    { inputs  :: [(TxIn, TxOut)]
      -- ^ Picked inputs
    , withdrawal :: Word64
      -- ^ An available withdrawal amount, counting as an extra input
    , reclaim :: Word64
      -- ^ Claim back a deposit, counting as a an extra input
    , outputs :: [TxOut]
      -- ^ Picked outputs
    , change  :: [Coin]
      -- ^ Resulting changes
    , deposit :: Word64
      -- ^ A deposit counting as an extra output
    } deriving (Generic, Show, Eq)
```

So, the presence and full value of a deposit can be inferred from a `CoinSelection`. Therefore, a better approach which does not duplicate any logic with regards to deposits would be to change `feeBalance` to return not only a `Fee`, but also a deposit value and have that bubbling through `estimateFeeForCoinSelection`.

Reply:
Doing it in `feeBalance` is quite intrusive and annoying. I believe it's best to do it in a lambda. So:

```hs
    estimateFeeForCoinSelection isKeyReg
        $ (\cs -> (Fee $ feeBalance cs, CoinSelection.deposit cs)) <$> selectCoins
```

Reply:
I think we should try again to move it out.